src/emmy/value.cljc

```diff
@@ -162,98 +142,33 @@
        java.lang.Float
        (numerical? [_] true)]))
 
-(extend-protocol Value
+(extend-protocol IKind
   #?(:clj Number :cljs number)
-  (zero? [x] (core/zero? x))
-  (one? [x] (== 1 x))
-  (identity? [x] (== 1 x))
-  (zero-like [_] 0)
-  (one-like [_] 1)
-  (identity-like [_] 1)
-  (freeze [x] x)
-  (exact? [x] #?(:clj  (or (integer? x) (ratio? x))
-                 :cljs (integer? x)))
   (kind [x] #?(:clj (type x)
-               :cljs (if (exact? x)
+               :cljs (if (and (. js/Number isInteger x)
+                              (< (Math/abs x) (.-MAX_SAFE_INTEGER js/Number)))
                        ::native-integral
                        ::floating-point)))
 
   #?(:clj Boolean :cljs boolean)
-  (zero? [_] false)
-  (one? [_] false)
-  (identity? [_] false)
-  (zero-like [_] 0)
-  (one-like [_] 1)
-  (identity-like [_] 1)
-  (freeze [x] x)
-  (exact? [_] false)
   (kind [x] (type x))
 
   #?@(:clj
       [java.lang.Double
-       (zero? [x] (core/zero? x))
-       (one? [x] (== 1 x))
-       (identity? [x] (== 1 x))
-       (zero-like [_] 0.0)
-       (one-like [_] 1.0)
-       (identity-like [_] 1.0)
-       (freeze [x] x)
-       (exact? [_] false)
        (kind [x] (type x))
 
        java.lang.Float
-       (zero? [x] (core/zero? x))
-       (one? [x] (== 1 x))
-       (identity? [x] (== 1 x))
-       (zero-like [_] 0.0)
-       (one-like [_] 1.0)
-       (identity-like [_] 1.0)
-       (freeze [x] x)
-       (exact? [_] false)
        (kind [x] (type x))])
 
   nil
-  (zero? [_] true)
-  (one? [_] false)
-  (identity? [_] false)
-  (zero-like [_] (u/unsupported "nil doesn't support zero-like."))
-  (one-like [_] (u/unsupported "nil doesn't support one-like."))
-  (identity-like [_] (u/unsupported "nil doesn't support identity-like."))
-  (freeze [_] nil)
-  (exact? [_] false)
   (kind [_] nil)
 
-  Var
-  (zero? [_] false)
-  (one? [_] false)
-  (identity? [_] false)
-  (zero-like [v] (u/unsupported (str "zero-like: " v)))
-  (one-like [v] (u/unsupported (str "one-like: " v)))
-  (identity-like [v] (u/unsupported (str "identity-like: " v)))
-  (freeze [v] (:name (meta v)))
-  (exact? [_] false)
-  (kind [v] (type v))
+;;  Var
```

Comment:
Currently we just have `kind` for nil (returning nil). Since nil no longer gets any of the other formerly-Value methods, I thought, should it have a `kind`?

If nil has no kind, i.e., it's not an Emmy value at all, a few things break. We lose this:

```
  (defmethod g/partial-derivative [t nil] [f _]
    (multivariate f [])))
```

Although I think that could be fixed by having the `(kind nil)` be `::v/seqtype`. There are a few other cases of error that we could fix.

But the experiment of removing `IKind` from `nil` has kind of convinced me that we want the `kind` function to be a total function, safe to call on any Clojure(Script) value, so that a defmethod could in principle be written for any object.

So I am comfortable with the status quo. Not having the other Value methods like `zero?` defends against unfortunate nil-puns.